examples/rust/event_multiplexing/wait.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2024 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Apache Software License 2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use clap::Parser;
+use iceoryx2::{port::listener::Listener, prelude::*};
+use std::collections::HashMap;
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let args = Args::parse();
+
+    let node = NodeBuilder::new().create::<ipc::Service>()?;
+
+    // factory lambda to create a listener with a given service name
+    let create_listener =
+        |service: &String| -> Result<Listener<ipc::Service>, Box<dyn std::error::Error>> {
+            let event = node
+                .service_builder(&service.as_str().try_into()?)
+                .event()
+                .open_or_create()?;
+
+            Ok(event.listener_builder().create()?)
+        };
+
+    let waitset = WaitSetBuilder::new().create::<ipc::Service>()?;
+    let mut listeners = vec![];
+    let mut listener_attachments: HashMap<
+        AttachmentId<ipc::Service>,
+        (&String, &Listener<ipc::Service>),
+    > = HashMap::new();
+    let mut guards = vec![];
+
+    // create a listener for every service
+    for service in &args.services {
+        listeners.push((service, create_listener(service)?));
+    }
+
+    // attach all listeners to the waitset and store the guard
+    for (service, listener) in &listeners {
+        let guard = waitset.attach_notification(listener)?;
+        listener_attachments.insert(guard.to_attachment_id(), (service, listener));
+        guards.push(guard);
+    }
+
+    println!("Waiting on the following services: {:?}", args.services);
+
+    // the callback that is called when a listener has received an event
+    let trigger_call = |attachment_id: AttachmentId<ipc::Service>| {
+        if let Some((service_name, listener)) = listener_attachments.get(&attachment_id) {
+            print!("Received trigger from \"{}\" ::", service_name);
+
+            while let Ok(Some(event_id)) = listener.try_wait_one() {
+                print!(" {:?}", event_id);
+            }
+
+            println!("");
+        }
+    };
+
+    // wait until at least one listener has received an event or the user has pressed CTRL+c
+    // or send SIGTERM/SIGINT
+    while waitset.run(trigger_call).is_ok() {}
```

Comment:
Just wondering...

1. Why name `run` instead of `wait`?
2. What is supposed to go in the `{}` ? Does there need to be a loop here or can it just be a function which blocks until either a timeout or `SIGINT` ?
3. Would it be cleaner to define the callback at this call?

```rust
waitset.wait(|attachment_id: AttachmentId<ipc::Service>| {
        if let Some((service_name, listener)) = listener_attachments.get(&attachment_id) {
            print!("Received trigger from \"{}\" ::", service_name);

            while let Ok(Some(event_id)) = listener.try_wait_one() {
                print!(" {:?}", event_id);
            }

            println!("");
        });
        
        // If we get here, either it timed-out (if timed wait) or there was a `SIGINT`.
```

Reply:
Regarding 1)
Further down there is an example with
```rs
while waitset.run(event_handler).is_ok() {}
```
That was a proposal which @elfenpiff liked so it was changed. The intention was to indicate that it's not just a blocking wait call but that actually something is executed.

Reply:
To me, having something like:

```rust
loop {
    waitset.wait(...); // or timed_wait
}
```

is more intuitive.

Reply:
1. @elBoberido wanted to have this name and I had no objections, I also have no objections to `wait`. Sounds like a duel between the two of you ... I will sit here and just watch ;)
2. Actually nothing. It is in the end the main loop of the program and the actual code goes into the callback. It looks a bit strange and the iceoryx1 waitset returned here a vector. This has the problem that we need to overallocate since we do not know beforehand how many events we received but with the callback approach we delegate this to the user.
3. Nope, this would produce a clippy warning. It is always in favor of declaring the lambda first and then provide it as an argument.

Reply:
No strong feelings about the name.
How would you call the thingie in the wait call?

Reply:
@elBoberido https://github.com/eclipse-iceoryx/iceoryx2/pull/428#discussion_r1784523618

Reply:
@orecham sorry, I don't get it. Keep in mind, it's already Friday evening. Don't expect too much cognitive capacity left 😅

Reply:
@elBoberido I mean, the argument, passed into the callback, is fine to be some ID, like the current `AttachmentId`.

The caller should track which IDs are which component they have attached, then depending on the id provided to the callback, react on the wake-up.

Reply:
Sorry, I was notclear enough. I didn't meant the parameter of the callback but the callback itself
```rs
fn foo(attachement_id: ...) {
    // ...
}
// ...
waitset.wait(foo);
```
How would you call `foo`?
Calling it just `callback` is similar to calling an `i64` just `integer`. Which is not wrong but also does not give any semantic information.
```rs
fn add(&mut self, integer: i64) {...}
// vs
fn add(&mut self, length: i64) {...}
```

Reply:
@elBoberido I didn't really want to name it anything, hence my suggestion (3) above, but this seems not feasible.

It's just a chunk of logic to execute when woken up by an event, so...

`on_wakeup` ? `on_event` ? `on_activation` ? `handler` ?

Reply:
I went with `on_event`.